File: backend/app/api/v1/invoices.py
from fastapi import APIRouter, Depends, Query, Response
from sqlalchemy.orm import Session
from typing import List
from app.core.database import get_db
from app.dependencies.auth import get_current_company
from app.schemas.common import ApiResponse
from app.schemas.invoice import InvoiceCreate, InvoiceResponse, InvoiceListResponse, InvoiceFinalizeRequest, InvoiceCancelRequest, InvoiceCalculateRequest, InvoiceCalculateResponse
from app.services.invoice_service import InvoiceService
from app.services.pdf_service import PdfService

router = APIRouter(prefix="/invoices", tags=["Invoices"])

@router.post("/calculate", response_model=ApiResponse[InvoiceCalculateResponse])
def calculate_invoice(request: InvoiceCalculateRequest, company = Depends(get_current_company), db: Session = Depends(get_db)):
    result = InvoiceService.calculate_invoice(db, str(company.id), company, request.model_dump())
    return ApiResponse(success=True, data=InvoiceCalculateResponse(**result))

@router.post("/", response_model=ApiResponse[InvoiceResponse])
def create_invoice(request: InvoiceCreate, company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoice = InvoiceService.create_invoice(db, str(company.id), company, request.model_dump())
    return ApiResponse(success=True, data=_invoice_to_response(invoice))

@router.get("/", response_model=ApiResponse[InvoiceListResponse])
def list_invoices(status: str = Query(None), company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoices = InvoiceService.list_invoices(db, str(company.id), status)
    return ApiResponse(success=True, data=InvoiceListResponse(
        items=[_invoice_to_response(i) for i in invoices],
        total=len(invoices)
    ))

@router.get("/{invoice_id}", response_model=ApiResponse[InvoiceResponse])
def get_invoice(invoice_id: str, company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoice = InvoiceService.get_invoice(db, str(company.id), invoice_id)
    return ApiResponse(success=True, data=_invoice_to_response(invoice))

@router.get("/{invoice_id}/pdf")
def get_invoice_pdf(invoice_id: str, company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoice = InvoiceService.get_invoice(db, str(company.id), invoice_id)
    pdf_bytes = PdfService.generate_invoice_pdf(invoice)
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f"attachment; filename=invoice_{invoice.invoice_number}.pdf"
        }
    )

@router.post("/{invoice_id}/finalize", response_model=ApiResponse[InvoiceResponse])
def finalize_invoice(invoice_id: str, request: InvoiceFinalizeRequest, company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoice = InvoiceService.finalize_invoice(db, str(company.id), invoice_id)
    return ApiResponse(success=True, data=_invoice_to_response(invoice))

@router.post("/{invoice_id}/cancel", response_model=ApiResponse[InvoiceResponse])
def cancel_invoice(invoice_id: str, request: InvoiceCancelRequest, company = Depends(get_current_company), db: Session = Depends(get_db)):
    invoice = InvoiceService.cancel_invoice(db, str(company.id), invoice_id, request.reason)
    return ApiResponse(success=True, data=_invoice_to_response(invoice))

def _invoice_to_response(invoice) -> InvoiceResponse:
    return InvoiceResponse(
        id=invoice.id,
        invoice_number=invoice.invoice_number,
        invoice_type=invoice.invoice_type,
        invoice_date=invoice.invoice_date.date(),
        customer_name_snapshot=invoice.customer_name_snapshot,
        customer_gstin_snapshot=invoice.customer_gstin_snapshot,
        place_of_supply=invoice.place_of_supply,
        subtotal=float(invoice.subtotal),
        discount_total=float(invoice.discount_total),
        taxable_total=float(invoice.taxable_total),
        cgst_total=float(invoice.cgst_total),
        sgst_total=float(invoice.sgst_total),
        igst_total=float(invoice.igst_total),
        grand_total=float(invoice.grand_total),
        amount_in_words=invoice.amount_in_words,
        invoice_status=invoice.invoice_status,
        payment_status=invoice.payment_status,
        notes=invoice.notes,
        lines=[{
            "id": l.id,
            "item_name": l.item_name_snapshot,
            "description": l.description_snapshot,
            "hsn_sac_snapshot": l.hsn_sac_snapshot,
            "quantity": float(l.quantity),
            "unit_name_snapshot": l.unit_name_snapshot,
            "unit_symbol_snapshot": l.unit_symbol_snapshot,
            "rate": float(l.rate),
            "discount_amount": float(l.discount_amount),
            "taxable_value": float(l.taxable_value),
            "gst_rate": float(l.gst_rate),
            "cgst_amount": float(l.cgst_amount),
            "sgst_amount": float(l.sgst_amount),
            "igst_amount": float(l.igst_amount),
            "line_total": float(l.line_total),
        } for l in invoice.lines],
        created_at=invoice.created_at,
    )

File: frontend/src/api/invoices.ts
import { apiClient } from './client';

export interface InvoiceLineCreate {
  item_id?: string | null;
  item_name: string;
  description?: string | null;
  hsn_sac?: string | null;
  quantity: number;
  unit_id: string;
  unit_name: string;
  unit_symbol: string;
  rate: number;
  discount_type?: string;
  discount_value?: number;
  gst_rate?: number;
}

export interface InvoiceCreateRequest {
  invoice_type?: string;
  invoice_date: string;
  customer_id: string;
  place_of_supply: string;
  lines: InvoiceLineCreate[];
  notes?: string | null;
  terms?: string | null;
}

export interface InvoiceCalculateRequest {
  customer_id?: string | null;
  place_of_supply: string;
  lines: InvoiceLineCreate[];
}

export interface InvoiceCalculateResponse {
  subtotal: number;
  discount_total: number;
  taxable_total: number;
  cgst_total: number;
  sgst_total: number;
  igst_total: number;
  grand_total: number;
  amount_in_words?: string | null;
  lines: any[];
}

export interface InvoiceResponse {
  id: string;
  invoice_number: string;
  invoice_type: string;
  invoice_date: string;
  customer_name_snapshot: string;
  place_of_supply: string;
  subtotal: number;
  discount_total: number;
  taxable_total: number;
  cgst_total: number;
  sgst_total: number;
  igst_total: number;
  grand_total: number;
  amount_in_words?: string | null;
  invoice_status: string;
  payment_status: string;
  notes?: string | null;
  lines: any[];
  created_at: string;
}

export const invoicesApi = {
  calculate: async (data: InvoiceCalculateRequest) => {
    const response = await apiClient.post<InvoiceCalculateResponse>('/invoices/calculate', data);
    return response.data;
  },
  create: async (data: InvoiceCreateRequest) => {
    const response = await apiClient.post<InvoiceResponse>('/invoices', data);
    return response.data;
  },
  getAll: async () => {
    const response = await apiClient.get<{items: InvoiceResponse[], total: number}>('/invoices');
    return response.data;
  },
  getById: async (id: string) => {
    const response = await apiClient.get<InvoiceResponse>(`/invoices/${id}`);
    return response.data;
  },
  finalize: async (id: string) => {
    const response = await apiClient.post<InvoiceResponse>(`/invoices/${id}/finalize`, {});
    return response.data;
  },
  cancel: async (id: string, reason: string) => {
    const response = await apiClient.post<InvoiceResponse>(`/invoices/${id}/cancel`, { cancel_reason: reason });
    return response.data;
  },
  getPdf: async (id: string) => {
    const response = await apiClient.get(`/invoices/${id}/pdf`, { responseType: 'blob' });
    const url = window.URL.createObjectURL(new Blob([response.data], { type: 'application/pdf' }));
    window.open(url, '_blank');
  }
};
